packages/eslint-plugin/src/rules/enum-utils/shared.ts

```diff
@@ -33,8 +50,40 @@ export function getEnumTypes(
   typeChecker: ts.TypeChecker,
   type: ts.Type,
 ): ts.Type[] {
-  return tsutils
-    .unionTypeParts(type)
-    .filter(subType => isTypeFlagSet(subType, ts.TypeFlags.EnumLiteral))
-    .map(type => getBaseEnumType(typeChecker, type));
+  return getEnumLiterals(type).map(type => getBaseEnumType(typeChecker, type));
+}
+
+/**
+ * Returns the enum key that matches the given literal node, or null if none
+ * match. For example:
+ * ```ts
+ * enum Fruit {
+ *   Apple = 'apple',
+ *   Banana = 'banana',
+ * }
+ *
+ * getEnumKeyForLiteral([Fruit.Apple, Fruit.Banana], 'apple') --> 'Fruit.Apple'
+ * getEnumKeyForLiteral([Fruit.Apple, Fruit.Banana], 'banana') --> 'Fruit.Banana'
+ * getEnumKeyForLiteral([Fruit.Apple, Fruit.Banana], 'cherry') --> null
+ * ```
+ */
+export function getEnumKeyForLiteral(
+  enumLiterals: ts.LiteralType[],
+  literal: unknown,
+): string | null {
+  for (const enumLiteral of enumLiterals) {
+    if (enumLiteral.value === literal) {
+      const { symbol } = enumLiteral;
+
+      const memberDeclaration = symbol.valueDeclaration as ts.EnumMember;
+      const enumDeclaration = memberDeclaration.parent;
+
+      const enumName = enumDeclaration.name.getText();
```

Comment:
[Cleanup] We generally try to avoid `getText()` when possible. It's a blunt function that directly draws from the source code text - which means it papers over nuances we might care about, such as if an `id` is actually a computed property.

```suggestion
      const enumName = enumDeclaration.name.text;
```

Reply:
Done